src/core/utils/components/VirtualScroll.tsx

```diff
@@ -0,0 +1,219 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import { CommonProps } from '../props';
+
+const getScrollableParent = (element: HTMLElement | null): HTMLElement => {
+  if (!element || element === document.body) {
+    return document.body;
+  }
+
+  return isElementScrollable(element)
+    ? element
+    : getScrollableParent(element.parentElement);
+};
+
+const isElementScrollable = (element: HTMLElement) => {
+  return /(auto|scroll|overlay)/.test(
+    getElementStyle(element, 'overflow') +
+      getElementStyle(element, 'overflow-y'),
+  );
+};
+
+const getElementStyle = (element: HTMLElement, prop: string) => {
+  return getComputedStyle(element, null).getPropertyValue(prop);
+};
+
+const getElementHeight = (element: HTMLElement | undefined) => {
+  return element?.getBoundingClientRect().height ?? 0;
+};
+
```

Comment:
If I'm not mistaken, parent container height changes are not watched. Container size could change for a variety of reasons, for example, due to user resizing a widget in iTwin.js.

Virtualization libraries usually rely on the component consumer to provide parent container height. This is often done to make libraries smaller and to avoid adding unnecessary ResizeObservers when parent height could be calculated by other means.

Reply:
How resizing widget would change elements container height? I mean, to have a scrollable container, you need to set its height, otherwise, scrollbar would not appear. So to change container height, user would subscribe to resizing event and manually change the css prop. Have you seen many use cases like that?  
Currently, VirtualScroll does not support body scroll (not sure, how valid it is too).

Reply:
itwin-viewer uses some virtualized components that fill whole widget area, i.e. resizing the widget alters the amount of visible rows. I don't see any way this could be implemented using the current VirtualScroll component without enormous waste.

You can see how the current implementation fails to handle dynamic container height by altering Virtualized Table story to set Table height to 100% and reducing bufferSize to 1. Then increase window height and observe how no items are being rendered below the initial fold.

Reply:
Added resize observer and listening to scroll container height changes. Whenever height changes, I recalculate virtual scroll.

Reply:
Thanks for adding support for dynamic content height.

However, I want to express a concern regarding the deviation from other windowing libraries, namely measuring parent height internally instead of receiving the height via a prop. Please review the arguments that I gave in my initial comment. 

Additionally, coming from our experience with iTwin.js virtualized components, people **will** forget to set their parent container height and come asking us why their component is not "working". Requiring to set the height prop explicitly solves this problem because the type checker will catch the error, and that's exactly what we've started doing since iTwin.js 3.0.

Reply:
well, if I ask for height prop, I should ask then for scroll container ref, which I was trying to add, but could not figure out a way (issue with ref/value). Cause it does not make sense to get the height, but not ask for the element. Would you have any ideas how I could do that? Then I could get rid of bunch of code to get the scollable container and then take its height.

I do not really get the "forgetting".
I mean, in native html and css, without setting the height, you would not have a scroll. And thats the way our components should also work. Rely on native things as much as possible without having some additional stuff on top of that.
It is not much we can do to help them learn. Developers need to have basic understanding of css or we will always end up in bugs and issues.
I would say then having extended documentation with examples how to achieve things and explain, why it needs to be done that way (which we kinda have with storybook, but also have Wiki and plan to have even separate website with everything at one place).